base/threadingconstructs.jl

```diff
@@ -3,24 +3,59 @@
 export threadid, nthreads, @threads, @spawn
 
 """
-    Threads.threadid()
+    Threads.threadid() -> Int
 
-Get the ID number of the current thread of execution. The master thread has ID `1`.
+Get the ID number of the current thread of execution within the current
+threadpool. The master thread has ID `1`.
 """
-threadid() = Int(ccall(:jl_threadid, Int16, ())+1)
+threadid() = Int(ccall(:jl_relative_threadid, Int16, ())+1)
 
-# Inclusive upper bound on threadid()
 """
-    Threads.nthreads()
+    Threads.threadid() -> Int
 
-Get the number of threads available to the Julia process. This is the inclusive upper bound
-on [`threadid()`](@ref).
+Get the ID number of the current thread of execution within the Julia session.
+The master thread has ID `1`.
+"""
+rawthreadid() = Int(ccall(:jl_threadid, Int16, ())+1)
+
+"""
+    Threads.nthreads(tp::Int=Threads.threadpoolid()) -> Int
+
+Get the number of threads available in the specified threadpool. This is the
+inclusive upper bound on [`threadid()`](@ref).
 
 See also: `BLAS.get_num_threads` and `BLAS.set_num_threads` in the
 [`LinearAlgebra`](@ref man-linalg) standard library, and `nprocs()` in the
 [`Distributed`](@ref man-distributed) standard library.
 """
-nthreads() = Int(unsafe_load(cglobal(:jl_n_threads, Cint)))
+nthreads(tp::Int=Threads.threadpoolid()) =
+    ccall(:jl_num_threads, Cint, (Cint,), tp-1)
+
+"""
+    Threads.spawn_threadpool(n::Int) -> Int
+
+Spawns a new threadpool of size `n`, and returns the threadpool ID.
```

Comment:
I'd suggest something like `(new|create|allocate)_threadpool` and return a `ThreadPool` object which can be passed to `@spawn`, `schedule`, `Channel`, etc.

Reply:
Marking as resolved since we no longer allow dynamically spawning threadpools.